### PR TITLE
AGS: Add GUIO options to disable autosaves/save management per-game

### PR DESCRIPTION
**Pull request by tag2015**
This long-overdue PR adds two flags that can be used in the detection tables to disable autosaves or completely disable save/loading from the GMM.

Many AGS games (mostly commercial) already feature an autosave mechanism, and also keeping enabled the scummvm
system causes save duplication and various annoyances (e.g. most Blackwell games autosave in the map screen using slot 0, causing the "found save in autosave slot" box to appear)

The option to disable saves entirely is for games that feature permadeath or similar modes (Dustbowl, Dreams in the Witch House, Whispers of a Machine...) to discourage cheating.

However, since some users may not appreciate the change, I also added a checkbox in game options to disable the feature.

The PR also includes a fix (which is more of a workaround to be honest) for the autosave that is shown twice in some Wadjet Eye games, as reported in [TRAC 13890](https://bugs.scummvm.org/ticket/13890).

NOTE: Some entries in the detection table still need to be updated. The PR is already mergeable, I'll finish the rest in the coming days

### CODE REVIEW

File: engines/ags/dialogs.cpp
Comment (Thunderforge) [inline engines/ags/dialogs.cpp:80]: What does "not fully compatible" mean in this context?

File: engines/ags/engine/ac/listbox.cpp
Comment (Thunderforge) [inline engines/ags/engine/ac/listbox.cpp:155]: Does this happen regardless of the override option?

**Review by Thunderforge — APPROVED** [review]

File: engines/ags/dialogs.cpp
Comment (tag2015) [reply engines/ags/dialogs.cpp:80]: Yeah that tooltip isn't too great, but I couldn't come up with something better... basically it means: "even for games that should manage saves (and autosaves) autonomously" (because ScummVM's features are disabled in the detection entry, that is)

File: engines/ags/engine/ac/listbox.cpp
Comment (tag2015) [reply engines/ags/engine/ac/listbox.cpp:155]: Yes, that is a separate fix that is applied in any case

File: engines/ags/dialogs.cpp
Comment (Thunderforge) [reply engines/ags/dialogs.cpp:80]: How about

```suggestion
	_overrideSavesCheckbox = new GUI::CheckboxWidget(widgetsBoss(), _dialogLayout + ".savesOvr", _("Always use ScummVM save management"), _("Do not allow games to use their built-in manual save and autosave management, and instead use ScummVM manual save and autosave management (NOT RECOMMENDED)"));
```

Also, why is it not recommended?

Comment (tag2015) [reply engines/ags/dialogs.cpp:80]: This sounds like the two options are mutually exclusive, but in reality it's:
- saves managed by the game, or
- saves managed by the game + scummvm
I have another proposal: "Never disable ScummVM save management, even for games that should manually handle saves and autosaves"
I could probably drop the "NOT RECOMMENDED", seems a bit too strong

Comment (Thunderforge) [reply engines/ags/dialogs.cpp:80]: > I have another proposal: "Never disable ScummVM save management, even for games that should manually handle saves and autosaves"

Could we just flip the option: "Disable ScummVM save management"?

Part of the confusion may be because this PR is addressing two issues:

1. Games like _The Shivah_ have in-game autosaves and ScummVM autosaves either duplicating effort or interfering with each other
2. Games like _Dustbowl_ or _Whispers of a Machine_ intentionally prevent manual saves, but ScummVM currently allows them

It seems to me like we might actually want two options:

1. Disable ScummVM auto-saving (prevents conflict with games that do their own autosaving)
2. Disable ScummVM manual saves (useful for games that allow permadeath, iron man, etc)

Comment (Thunderforge) [reply engines/ags/dialogs.cpp:80]: For what it's worth, here's a summary of how various games work in the original interpreter:

Game | In-Game Manual Saves | In-Game Auto-Saves
--- | --- | --- 
_Steamed Hams_ | ❌ | ❌
Typical AGS game? | ✅ | ❌
_Whispers of a Machine_ | ❌ | ✅
_The Blackwell Legacy_ | ✅ | ✅ 

And here's how they currently work in ScummVM:

Game | In-Game Manual Saves | ScummVM Manual Saves | In-Game Auto-Saves | ScummVM Auto-Saves | Notes
--- | --- | --- | --- | --- | ---
_Steamed Hams_ | ❌ | ✅ | ❌ | ✅ | Extra save options are nice
Typical AGS game? | ✅ | ✅ | ❌ | ✅ | Manual saves either way are fine, extra autosaves are nice
_Dustbowl_ | ❌ | ✅ | ✅ | ✅ | Manual saves are against the spirit of the game, but may be desired. In-game and ScummVM autosaves interfere
_The Blackwell Legacy_ | ✅ | ✅ | ✅ | ✅ | In-game and ScummVM autosaves interfere

Note that for some games it's a good thing to have ScummVM allowing manual saves when it wasn't offered and for some games it's not, and the same with autosaves.

Comment (tag2015) [reply engines/ags/dialogs.cpp:80]: Not sure about this... I wanted to keep things simple from a user point of view:
- If you leave the checkbox unchecked, ScummVM uses the settings setup internally (which _may_ disable the GMM saves or autosaves)
- If you check the box, all the usual features (autosave and GMM saving) are always enabled, but you take the chance that there may be inconsistencies

If we decide to use two "disable ..." checkboxes, besides the fact that the entire thing works in reverse (that means that the boxes must be automatically ticked on addition of e.g. Dustbowl, or the user wouldn't bother changing the option)
it could bring additional confusion:
Let's say that the uses ticks "disable scummvm manual saves" for a generic game:
- if we silently ignore the option, it seems the checkbox has no effect in general,
- if we enforce the option, it just disables a perfectly working functionality to no gain

Also, these settings assume that the user knows well how the games and scummvm handle saves, which imho shouldn't be taken for granted...
I'd rather wait for more feedback before going one way or the other

Comment (bluegr) [reply engines/ags/dialogs.cpp:80]: The handling of autosaves isn't something that should be decided by the users. This should be handled by the engine itself, enforcing either the ScummVM semantics or the game ones. IMHO, games that handle autosaving on their own should have an option in the detection entry, which will always override and disable the in-built ScummVM autosave functionality.

Regarding the regular save functionality: I believe it's superfluous to have a checkbox for that. Since the save/load menu is a global functionality, there is no point to disable it with a checkbox: users may access it if they wish, which is cheating in some games, but it's their choice. If they want to avoid cheating, they don't need an extra checkbox to ensure that they won't cheat :)

Comment (tag2015) [reply engines/ags/dialogs.cpp:80]: Disabling GMM saving is not only to prevent cheating but also to prevent using the functionality in games where it doesn't work properly (such as Strangeland), so the player has to use the game menu instead of scummvm's.

I agree with your first point regarding autosaves, however I decided to add the checkbox basically to keep the current functionality... ticking it disables this PR's effects :)
All things considered i think that a single checkbox is the cleanest implementation.

Comment (Thunderforge) [reply engines/ags/dialogs.cpp:80]: > I agree with your first point regarding autosaves, however I decided to add the checkbox basically to keep the current functionality...

Are there situations where we _would_ want the current functionality? I'm thinking for games like _The Blackwell Legacy_ and _The Shivah_ where there are double autosaves, I don't think the current behavior would ever be desired.

Comment (tag2015) [reply engines/ags/dialogs.cpp:80]: Regardless of the checkbox, this PR is temporarily on hold until PR #5496 is merged (since it uses systemwide functions that are directly touched by that PR)

**Comment by tag2015** [comment]
Added the custom message, and tweaked the checkbox a bit.
Now when attempting to save/load in a game with GMM disabled, this popup is shown
warning

And this is the override checkbox
checkboxtrim

**Comment by tag2015** [comment]
I hoped to get this ready for 2.8.0 but it's too late now. I merged (and I'll backport) commit `d2de01b303c286d5ef870f7202ba54322ba3c06f`; for the rest we'll have time to decide

File: engines/ags/dialogs.cpp
Comment (criezy) [inline engines/ags/dialogs.cpp:81]:
```suggestion
	// Override game save management
	if (Common::checkGameGUIOption(GAMEOPTION_NO_AUTOSAVE, ConfMan.get("guioptions", domain)) ||
		Common::checkGameGUIOption(GAMEOPTION_NO_SAVELOAD, ConfMan.get("guioptions"))) {
		_overrideSavesCheckbox = new GUI::CheckboxWidget(widgetsBoss(), _dialogLayout + ".savesOvr", _("Force ScummVM save management"), _("Never disable ScummVM save management and autosaves.\nNOTE: This could cause save duplication and other oddities"));
	else
		_overrideSavesCheckbox  = nullptr;

```
And add a sanity check where `_overrideSavesCheckbox` is used to avoid dereferencing a null pointer.

I assume the checkbox only make sense for games that have those flags?

**Comment by criezy** [comment]
I like the change in this pull request. I only added a suggestion to only show the checkbox for games that have the relevant flags.

**Comment by tag2015** [comment]
@criezy I added the change you requested if you want to give it a test, I think the PR can be merged

**Comment by criezy** [comment]
It looks good to me. My only nitpick is with the label of the button. `Force ScummVM save management` may give the impression we are forcing the user to use it instead of the original game save mechanism. Maybe `Enable ScummVM save management` would be better, especially now that the button is only shown for games that disable it by default?

**Comment by tag2015** [comment]
> It looks good to me. My only nitpick is with the label of the button. `Force ScummVM save management` may give the impression we are forcing the user to use it instead of the original game save mechanism. Maybe `Enable ScummVM save management` would be better, especially now that the button is only shown for games that disable it by default?

Hm, you're right. I used "force" for consistency with the other checkbox but `Enable ScummVM save management` may be better. I'll change that

**Comment by criezy** [comment]
Thank you

**Pull request merged**
